3rdparty/everest/library/x25519.c

```diff
@@ -0,0 +1,187 @@
+                        void *p_rng )
+{
+    int ret = 0;
+
+    uint8_t base[32] = {0};
+
+    if( ( ret = f_rng( p_rng, ctx->our_secret, 32 ) ) != 0 )
+        return ret;
+
+    *olen = 36;
+    if( blen < *olen )
+        return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+
+    *buf++ = MBEDTLS_ECP_TLS_NAMED_CURVE;
+    *buf++ = MBEDTLS_ECP_TLS_CURVE25519 >> 8;
+    *buf++ = MBEDTLS_ECP_TLS_CURVE25519 & 0xFF;
+    *buf++ = 32;
+
+    base[0] = 9;
+    Hacl_Curve25519_crypto_scalarmult( buf, ctx->our_secret, base );
+
+    base[0] = 0;
+    if( memcmp( buf, base, 32) == 0 )
+        return MBEDTLS_ERR_ECP_RANDOM_FAILED;
+
+    return( 0 );
+}
+
+int mbedtls_x25519_read_params( mbedtls_x25519_context *ctx,
+                        const unsigned char **buf, const unsigned char *end )
+{
+    if( end - *buf < 33 )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    if( ( *(*buf)++ != 32 ) )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    memcpy( ctx->peer_point, *buf, 32 );
+    *buf += 32;
+    return( 0 );
+}
+
+int mbedtls_x25519_get_params( mbedtls_x25519_context *ctx, const mbedtls_ecp_keypair *key,
+                        int side )
+{
+    size_t olen = 0;
+
+    switch( side ) {
+    case MBEDTLS_ECDH_THEIRS:
+        mbedtls_ecp_point_write_binary( &key->grp, &key->Q, MBEDTLS_ECP_PF_COMPRESSED, &olen, ctx->peer_point, 32 );
+        /* untested; defensively throw an error for now. */
+        return(MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE);
+    case MBEDTLS_ECDH_OURS:
+        mbedtls_mpi_write_binary( &key->d, ctx->our_secret, 32 );
+        /* CMW: key->Q = key->d * base; do we need to set up ctx.peer_point here? */
+        /* untested; defensively throw an error for now. */
+        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+    default:
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+    }
+}
+
+int mbedtls_x25519_calc_secret( mbedtls_x25519_context *ctx, size_t *olen,
+                        unsigned char *buf, size_t blen,
+                        int( *f_rng )(void *, unsigned char *, size_t),
+                        void *p_rng )
+{
+    /* CMW: Is it okay that f_rng, p_rng are not used? */
+    (( void )f_rng);
+    (( void )p_rng);
+
+    *olen = 32;
+
+    if( blen < *olen )
+        return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+
+    Hacl_Curve25519_crypto_scalarmult( buf, ctx->our_secret, ctx->peer_point);
+
+    /* Wipe the DH secret and don't let the peer chose a small subgroup point */
+    memset( ctx->our_secret, 0, 32 );
+    if( memcmp( buf, ctx->our_secret, 32) == 0 )
+        return MBEDTLS_ERR_ECP_RANDOM_FAILED;
+
+    return( 0 );
+}
+
+int mbedtls_x25519_make_public( mbedtls_x25519_context *ctx, size_t *olen,
+                        unsigned char *buf, size_t blen,
+                        int( *f_rng )(void *, unsigned char *, size_t),
+                        void *p_rng )
+{
+    unsigned char base[32] = { 0 };
+
+    /* CMW: Is it okay that f_rng, p_rng are not used? */
```

Comment:
Good catch, thanks! The function names (all of ECDH) are really confusing and I have to go and read the documentation from scratch every single time I look at one of them. Fixed in 75879c150b377f27aa0d3612b7b58bd864eff3cc. This was also the reason for why I needed `*_gen_public` in `benchmark.c`, because `*_gen_public` adds a private key to the context (obviously!).